File: peek_core_device/_private/storage/DeviceInfoTable.py
import logging

from peek_core_device.tuples.DeviceInfoTuple import DeviceInfoTuple
from sqlalchemy import Column
from sqlalchemy.sql.sqltypes import Boolean
from sqlalchemy.sql.sqltypes import DateTime
from sqlalchemy.sql.sqltypes import Integer
from sqlalchemy.sql.sqltypes import String

from .DeclarativeBase import DeclarativeBase
from ...tuples.DeviceGpsLocationTuple import DeviceGpsLocationTuple

logger = logging.getLogger(__name__)


class DeviceInfoTable(DeclarativeBase):
    """DeviceInfoTable

    This table stores information about devices.

    """

    __tablename__ = "DeviceInfo"

    TYPE_MOBILE_IOS = "mobile-ios"
    TYPE_MOBILE_ANDROUD = "mobile-android"
    TYPE_MOBILE_WEB = "mobile-web"
    TYPE_DESKTOP_WEB = "desktop-web"
    TYPE_DESKTOP_WINDOWS = "desktop-windows"
    TYPE_DESKTOP_MACOS = "desktop-macos"

    id = Column(Integer, primary_key=True)
    description = Column(String(100), nullable=False, unique=True)
    deviceId = Column(String(50), nullable=False, unique=True)
    deviceType = Column(String(20), nullable=False)
    deviceToken = Column(String(50), nullable=False, unique=True)
    appVersion = Column(String(15), nullable=False)
    updateVersion = Column(String(15))  # Null means it hasn't updated
    lastOnline = Column(DateTime(True))
    lastUpdateCheck = Column(DateTime(True))
    createdDate = Column(DateTime(True), nullable=False)
    deviceStatus = Column(Integer, nullable=False, server_default="0")
    isEnrolled = Column(Boolean, nullable=False, server_default="0")

    def toTuple(
        self,
        currentLocationTuple: DeviceGpsLocationTuple = None,
    ):
        return self.toTupleStatic(self, currentLocationTuple)

    @staticmethod
    def toTupleStatic(
        table: "DeviceInfoTable",
        currentLocationTuple: DeviceGpsLocationTuple = None
    ):
        return DeviceInfoTuple(
            description=table.description,
            deviceId=table.deviceId,
            deviceType=table.deviceType,
            deviceToken=table.deviceToken,
            appVersion=table.appVersion,
            updateVersion=table.updateVersion,
            lastOnline=table.lastOnline,
            lastUpdateCheck=table.lastUpdateCheck,
            createdDate=table.createdDate,
            deviceStatus=table.deviceStatus,
            isEnrolled=table.isEnrolled,
            currentLocation=currentLocationTuple,
        )

File: peek_core_device/_private/admin-app/tuples/DeviceInfoTable.ts
import { addTupleType, Tuple } from "@synerty/vortexjs"
import { DeviceTypeEnum, deviceTuplePrefix } from "@peek/peek_core_device/_private"
import { DeviceGpsLocationTuple } from "@peek/peek_core_device"
import { Capacitor } from "@capacitor/core"

@addTupleType
export class DeviceInfoTable extends Tuple {
    public static readonly tupleName = deviceTuplePrefix + "DeviceInfoTable"

    readonly TYPE_FIELD_IOS = "field-ios"
    readonly TYPE_FIELD_ANDROID = "field-android"
    readonly TYPE_MOBILE_WEB = "mobile-web"
    readonly TYPE_DESKTOP_WEB = "desktop-web"
    readonly TYPE_DESKTOP_WINDOWS = "desktop-windows"
    readonly TYPE_DESKTOP_MACOS = "desktop-macos"
    
    readonly DEVICE_OFFLINE = 0
    readonly DEVICE_ONLINE = 1
    readonly DEVICE_BACKGROUND = 2

    id: string
    description: string
    deviceId: string
    deviceType: string
    deviceToken: string
    appVersion: string
    updateVersion: string
    lastOnline: Date
    lastUpdateCheck: Date
    createdDate: Date
    deviceStatus: number
    isEnrolled: boolean
    currentLocation: DeviceGpsLocationTuple

    constructor() {
        super(DeviceInfoTable.tupleName)
    }

    get isWeb(): boolean {
        return !Capacitor.isNative
    }

    get isBackgrounded(): boolean {
        return !!(this.deviceStatus & this.DEVICE_BACKGROUND)
    }

    get googleMapLink() {
        if (!this.hasCurrentLocation()) {
            throw new Error("current location is not available")
        }
        return "https://www.google.com/maps/search/?api=1&query="
            + `${this.currentLocation.latitude},${this.currentLocation.longitude}`
    }

    setDeviceType(val: DeviceTypeEnum) {
        switch (val) {
            case DeviceTypeEnum.DESKTOP_WEB:
                this.deviceType = this.TYPE_DESKTOP_WEB
                break

            case DeviceTypeEnum.DESKTOP_MACOS:
                this.deviceType = this.TYPE_DESKTOP_MACOS
                break

            case DeviceTypeEnum.DESKTOP_WINDOWS:
                this.deviceType = this.TYPE_DESKTOP_WINDOWS
                break

            case DeviceTypeEnum.FIELD_IOS:
                this.deviceType = this.TYPE_FIELD_IOS
                break

            case DeviceTypeEnum.FIELD_ANDROID:
                this.deviceType = this.TYPE_FIELD_ANDROID
                break

            case DeviceTypeEnum.MOBILE_WEB:
                this.deviceType = this.TYPE_MOBILE_WEB
                break

        }
    }

    hasCurrentLocation() {
        if (!this.currentLocation?.latitude) {
            return false
        }
        if (typeof this.currentLocation.latitude === "number") {
            return true
        }
        return false
    }
}
